Enable string conversion in EUC-JP.

YOCKOW opened the pull request:
Background: EUC-JP is not supported by OSS CoreFoundation, while it is supported by macOS Foundation Framework.
See https://github.com/swiftlang/swift-foundation/issues/1016

This PR resolves the issue by calling ICU API if necessary.

YOCKOW commented:
@swift-ci Please test

parkera commented on Sources/FoundationEssentials/String/StringProtocol+Essentials.swift:
```diff
@@ -255,8 +260,12 @@ extension String {
+                // Attempt an up-call into swift-corelibs-foundation, which can defer to the CoreFoundation implementation
+                _cfStringEncodingConvert(string: self, using: encoding.rawValue, allowLossyConversion: allowLossyConversion) ??
+                // Or attempt an up-call into ICU via FoundationInternationalization
+                _icuStringEncodingConvert(string: self, using: encoding, allowLossyConversion: allowLossyConversion)
```
When do we fall back to the ICU one here? On Darwin this encoding is handled via CF, which itself calls into ICU.

parkera replied:
I guess that'll happen if `import Foundation` has not been done (and CF is present), but `import FoundationInternationalization` has. I wonder how we can further unify these paths; on Darwin too...

jmschonfeld replied:
I had a similar question - I believe `_cfStringEncodingConvert` will return `nil` if any of the following are the case:

1. CF isn't loaded
2. CF is loaded but it doesn't support the provided encoding
3. CF is loaded and supports the encoding but the data is malformed

We want the up-call in 2, and I think Tony's question was about 1, but do we want the up call in case 3?

YOCKOW replied:
I now agree that unexpected fallback should be avoided.
Although ICU *can* support other encodings, I shoud've make this change focus on EUC-JP.

parkera commented on Sources/FoundationInternationalization/ICU/ICU+StringConverter.swift:
```diff
@@ -0,0 +1,202 @@
+internal import _FoundationICU
+
+private extension String.Encoding {
+    var _icuConverterName: String? {
```
Should we make sure this returns `nil` for any encoding which we never want to use ICU for? `UTF8`, `ASCII`, `UTF16/32`, etc.

YOCKOW replied:
Certainly.
I'll let them return `nil` here or in `ICU.StringConverter`'s `init`.

jmschonfeld commented on Sources/FoundationInternationalization/ICU/ICU+StringConverter.swift:
```diff
@@ -0,0 +1,202 @@
+}
+
+extension ICU {
+    final class StringConverter: @unchecked Sendable {
```
nit: does this need the `@unchecked`? I would have assumed that since it's a `final` class with all immutable, `Sendable` properties that the compiler can validate this conformance

YOCKOW replied:
Thank you.
That's a vestige of my first implementation where `_converter` was a bare pointer...

YOCKOW replied:
Sorry.
`@unchecked` is necessary still because `LockedState` is not `Sendable`...

jmschonfeld replied:
Ah `LockedState` is `Sendable` but it requires its stored value to be of a `Sendable` type, and I had mistakenly thought `OpaquePointer` was `Sendable` but it is indeed not `Sendable` which is why the `Sendable` conformance isn't available there. This `@unchecked Sendable` makes sense then, since despite the pointer and anything it stores never escapes the lock.

jmschonfeld commented on Sources/FoundationInternationalization/ICU/ICU+StringConverter.swift:
```diff
@@ -0,0 +1,202 @@
+}
+
+extension ICU.StringConverter {
+    nonisolated(unsafe) static private var _converters: LockedState<[String.Encoding: ICU.StringConverter]> = .init(initialState: [:])
```
I think this should just be a `static private let` (the immutable-ness affects the `LockedState` itself but not the contents IIRC, and this isn't unsafe because it's guarded by a lock)

YOCKOW replied:
Indeed it should be.
This is also a vestige of my first impl, I guess.

jmschonfeld commented on Sources/FoundationInternationalization/ICU/ICU+StringConverter.swift:
```diff
@@ -0,0 +1,202 @@
+        return nil
+    }
+    let data =  Data(
+        bytesNoCopy: UnsafeMutableRawPointer(mutating: pointer),
```
Is it worth adding a comment here about this being "safe" because this data neither escapes this function nor is mutated? Since the pointer provided is not guaranteed to live longer than the function or be mutable, this could be a trap for a later addition that doesn't see those two constraints

YOCKOW replied:
I agree it's better to have some comments. I'm going to add ones.

YOCKOW commented:
@swift-ci Please test

YOCKOW commented:
@parkera @jmschonfeld Thank you for reviewing.

In response to your feedback, I've made some changes:
- Made it clear that `ICU.StringConverter` doesn't support `UTF-*` and `US-ASCII`.
- Removed unnecessary `nonisolated(unsafe)` from the static property.
- Added comments in `func _icuMakeStringFromBytes_impl` to clarify the safety of passing the pointer.
- Let it delegate string conversion to ICU only if the encoding is `EUC-JP` and there's no `FOUNDATION_FRAMEWORK` (at this point).

YOCKOW commented:
Please let me request reviews again as per changes above.

jmschonfeld commented on Sources/FoundationEssentials/String/String+IO.swift:
```diff
@@ -184,6 +189,17 @@ extension String {
+            // but are not supported by corelibs-foundation.
+            // We delegate conversion to ICU.
+            guard let string = (
+                bytes.withContiguousStorageIfAvailable({ _icuMakeStringFromBytes($0, encoding: encoding) }) ??
```
Actually sorry one more thing that I just caught - this `_icuMakeStringFromBytes` function is only defined in `!FOUNDATION_FRAMEWORK` builds so this fails to build with `FOUNDATION_FRAMEWORK` enabled. Could you wrap this case in `#if !FOUNDATION_FRAMEWORK`. For now in Foundation.framework this will fall back to `NSString` below, and we can look at using ICU instead of `NSString` in Foundation.framework in the future (since that may be a behavioral change. Same goes for the case in `StringProtocol+Essentials.swift` as well

YOCKOW replied:
Thank you for the review. This is indeed wrapped by `#if !FOUNDATION_FRAMEWORK` already:


```swift
#if !FOUNDATION_FRAMEWORK
case .isoLatin1:
  ...
case .macOSRoman:
  ...
case .japaneseEUC:
  ...
#endif
```

jmschonfeld commented on Sources/FoundationInternationalization/ICU/ICU+StringConverter.swift:
```diff
@@ -0,0 +1,204 @@
+}
+
+
+@_dynamicReplacement(for: _icuMakeStringFromBytes(_:encoding:))
```
This replaced function is only defined in `!FOUNDATION_FRAMEWORK` builds, could we wrap this in `#if !FOUNDATION_FRAMEWORK` as well so that it will build successfully in `FOUNDATION_FRAMEWORK` mode?

YOCKOW replied:
Done!

jmschonfeld commented on Tests/FoundationInternationalizationTests/StringTests+Data.swift:
```diff
@@ -0,0 +1,126 @@
+        XCTAssertNil(sushi.data(using: String._Encoding.japaneseEUC))
+        XCTAssertEqual(
+            sushi.data(using: String._Encoding.japaneseEUC, allowLossyConversion: true),
+            "Sushi?".data(using: .utf8)
```
This test unfortunately fails in the `FOUNDATION_FRAMEWORK` configuration because `Foundation.framework`'s implementation of this conversion replaces both UTF-16 scalars of the emoji with `?` rather than replacing the entire grapheme cluster with a single `?`. I suspect that maybe ICU is replacing only with a single `?` - do you happen to know if that's intentional? If so, for now we might need to vary the expected value to be `Sushi??` in `FOUNDATION_FRAMEWORK` vs. `Sushi?` in `!FOUNDATION_FRAMEWORK` to account for the discrepancy between `Foundation.framework` and ICU (assuming we think this result from ICU is indeed correct and not a bug in how we're calling ICU)

jmschonfeld replied:
Yeah I suspect this behavior comes from the fact that `NSString` is natively stored as UTF-16. For now, let's just divide the test case with a `FOUNDATION_FRAMEWORK`/`!FOUNDATION_FRAMEWORK` value for this expectation since I think enabling the conversion at all on non-Darwin is a good step even if the behavior is slightly different than `Foundation.framework`, and I can come back with a separate change to look into making this behavior the same across both platforms (while ensuring we don't break compatibility with existing clients).

jmschonfeld reviewed (commented):
Thanks - that resolves the build failures, but I have a comment about one issue in the tests that I've discovered as well (this should hopefully be the last thing, thanks for bearing with me)

YOCKOW commented:
@jmschonfeld
Thank you for the advice. I revised the test cases.

I didn't know CI didn't execute test cases with `FOUNDATION_FRAMEWORK`.

YOCKOW commented:
@swift-ci Please test
